src/EVM/SMT.hs

```diff
@@ -103,8 +104,14 @@ data SMTCex = SMTCex
 
 
 -- | Used for abstraction-refinement of the SMT formula. Contains assertions that make our query fully precise. These will be added to the assertion stack if we get `sat` with the abstracted query.
-newtype RefinementEqs = RefinementEqs [Builder]
-  deriving (Eq, Show, Monoid, Semigroup)
+data RefinementEqs = RefinementEqs [Builder] [Prop]
```

Comment:
what are the extra props for?

Reply:
So the SMT2 can store the original set of `Prop`-s that were used to generate the SMT2 query. When the SMT2 gets to `runTask`, it only has the SMT2. I wanted the SMT2 to also contain the `Prop`-s that were used to create it, so we can dump the props together with the SMT2 query when we actually send it to the solver.

Honestly, without a facility like this PR, it is essentially impossible for me to do proper debugging. I think you probably have other methods of doing this, but I have had a bit of a trouble fighting the repl just trying to get something meaningful info out of a single test file. I'd rather have something that actually works by simply setting a debug flag, without me having to understand the full, recursive, function-passing, iterative, self-folding callchain and insert some helper function/trace every time I need to do some minor debugging.

Reply:
Note that having the `Prop`-s at the top of the SMT2 file is incredibly valuable. I wanted to make sure that information stays there, and is always exactly correct. The only way I found to ensure this is to pass it along with the SMT2, inserting it as part of `assertProps`. This is the only way I have found that can do that while making sure things stay relatively clean & always having matching `Prop`-s to the SMT2 being generated.